:sparkles: Support results output as in-toto statement

puerco opened the pull request:
#### What kind of change does this PR introduce?

Feature

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

Currently scorecard does not support generating in-toto statements.

#### What is the new behavior (if this is a feature change)?**

This PR adds support to generate the scorecard output in an in-toto statement. It introduces a new possible value for the format flag: `--format=intoto`.

**The predicate format is open to discussion**, right now is just a reformated version of `JSONScorecardResultV2` with the repo removed (as it is now in the statement's subject section):

```
scorecard --format=intoto --checks=Maintained --repo=github.com/protobom/protobom
```

```json
{
  "type": "https://in-toto.io/Statement/v1",
  "subject": [
    {
      "name": "github.com/protobom/protobom",
      "uri": "git+https://github.com/protobom/protobom@0aa9659c6dc7354c5c60945f44b35bb2ed457857",
      "digest": {
        "gitCommit": "0aa9659c6dc7354c5c60945f44b35bb2ed457857"
      }
    }
  ],
  "predicate_type": "https://scorecard.dev/result/v0.1",
  "predicate": {
    "date": "2025-02-19T11:32:25-06:00",
    "scorecard": {
      "version": "devel",
      "commit": "unknown"
    },
    "score": 10.0,
    "checks": [
      {
        "details": null,
        "score": 10,
        "reason": "30 commit(s) and 5 issue activity found in the last 90 days -- score normalized to 10",
        "name": "Maintained",
        "documentation": {
          "url": "https://github.com/ossf/scorecard/blob/main/docs/checks.md#maintained",
          "short": "Determines if the project is \"actively maintained\"."
        }
      }
    ],
    "metadata": null
  }
}
```

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Fixes #3352

#### Special notes for your reviewer

Let me know if the predicate format should change (or if fields should be added or removed).

/cc @adityasaky as the original author of #3352
/cc @marcelamelara @mlieberman85  

#### Does this PR introduce a user-facing change?

```release-note
Scorecard can now generate its output as an in-toto statement by specifying --format=intoto 
```

adityasaky commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+const (
+	PredicateType = "https://scorecard.dev/Result/v1"
```
```suggestion
	PredicateType = "https://scorecard.dev/result/v1"
```

there's no in-toto specific convention (AFAIK) for this to be capitalized. I know Statement is but I think most predicate types don't capitalize.

puerco replied:
OK, sounds good, changed.

adityasaky commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,82 @@
+func (r *Result) AsStatement(writer io.Writer, checkDocs docs.Doc, opt *AsJSON2ResultOption) error {
+	// Attestatio subject
+	subject := intoto.ResourceDescriptor{
+		Name: r.Repo.Name,
```
should this set the URI field by any chance?

puerco replied:
OK, it now populates the URi with an spdx-style downlaod locator (check the updated example above)

adityasaky commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,82 @@
+	subject := intoto.ResourceDescriptor{
+		Name: r.Repo.Name,
+		Digest: map[string]string{
+			"sha1": r.Repo.CommitSHA,
```
in-toto/attestation supports `gitCommit` so that's an option here. https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md#gitcommit-gittree-gitblob-gittag

puerco replied:
Done :+1:

adityasaky reviewed (commented):
This is neat, @puerco!

puerco commented:
OK, I've updated the patch with changes addressing the statement suggestions

spencerschrock commented on pkg/scorecard/json.go:
```diff
@@ -160,10 +159,10 @@ func (r *Result) AsJSON2(writer io.Writer, checkDocs docs.Doc, opt *AsJSON2Resul
 		if e != nil {
-			return sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("GetCheck: %s: %v", checkResult.Name, e))
+			return out, fmt.Errorf("GetCheck: %s: %w", checkResult.Name, e)
 		}
 		if doc == nil {
-			return sce.WithMessage(sce.ErrScorecardInternal, fmt.Sprintf("GetCheck: %s: %v", checkResult.Name, errNoDoc))
+			return out, fmt.Errorf("GetCheck: %s: %w", checkResult.Name, errNoDoc)
```
changing the error return like this is a breaking change here if any library callers are checking for `ErrScorecardInternal`.

spencerschrock replied:
Oops, forgot you made this a helper, it's still a `ErrScorecardInternal` for `AsJSON2`, so no breaking change here.

spencerschrock commented on options/options.go:
```diff
@@ -88,6 +88,8 @@ const (
 	FormatDefault = "default"
 	// FormatRaw specifies that results should be output in raw format.
 	FormatRaw = "raw"
+	// FormatStatement specifies that results should be output in an in-toto statement.
+	FormatStatement = "statement"
```
discussion: any preference on `statement` vs `intoto` ?

puerco replied:
I also considered `attestation`.

marcelamelara replied:
I think `intoto` would make it crystal clear what the output format would be.

spencerschrock replied:
Let's do `FormatIntoto = "intoto"`, I'll leave a few other comments on names.

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,83 @@
+	Repo *jsonRepoV2 `json:"repo,omitempty"`
+	JSONScorecardResultV2
+}
+
+// AsStatement converts the results as an int-toto statement.
```
typo: `int-toto`

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,83 @@
+}
+
+// AsStatement converts the results as an int-toto statement.
+func (r *Result) AsStatement(writer io.Writer, checkDocs docs.Doc, opt *AsJSON2ResultOption) error {
+	// Attestatio subject
```
do we envision any configuration we'd want to do independent of the JSON where we should we define an AsStatementOption (or `AsIntotoOption` depending on other comment)?

(also: typo in `Attestation`)

puerco replied:
Good idea. Right now, there is nothing to change, but I think we'll iterate the predicate or add other types, so specifying the predicate type will probably be the first option. I've wrapped the JSON2 options in a new AsStatement options set (which is empty for now).

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,83 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Subject {
+// 	 Repo           jsonRepoV2          `json:"repo"`
+// }
```
is this intentionally still in here? or leftover?

puerco replied:
Oops I didn't see that with the copyright boilerplate next to it.. Removed

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,83 @@
+	sce "github.com/ossf/scorecard/v5/errors"
+)
+
+const (
+	PredicateType = "https://scorecard.dev/result/v1"
+)
```
does `v1` imply predicate stability? Should this be v0.1 like in this comment?
https://github.com/ossf/scorecard/issues/3352#issuecomment-2368655593

puerco replied:
Yes, I agree. I switched it to 0.1 :+1:

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,95 @@
+	"github.com/ossf/scorecard/v5/log"
+)
+
+const (
+	PredicateType = "https://scorecard.dev/result/v0.1"
+)
+
+type Statement struct {
+	intoto.Statement
+	Predicate Predicate `json:"predicate"`
+}
+
+// Predicate overrides JSONScorecardResultV2 with a nullable Repo field.
+type Predicate struct {
+	Repo *jsonRepoV2 `json:"repo,omitempty"`
+	JSONScorecardResultV2
+}
```
do we expect someone to reference these? Or in general, do we need to export anything here other than `AsIntoto` and `AsIntotoResultOption`?

puerco replied:
I think other tools that want to parse the scorecard predicate could use the type to decode it. I've pushed another commit hidding the statement as it is not really useful.

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,95 @@
+	JSONScorecardResultV2
+}
+
+// AsStatementResultOption wraps AsJSON2ResultOption preparing it for.
+type AsStatementResultOption struct {
```
`AsIntotoResultOption`

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,95 @@
+	AsJSON2ResultOption
+}
+
+// AsStatement converts the results as an in-toto statement.
+func (r *Result) AsStatement(writer io.Writer, checkDocs docs.Doc, opt *AsStatementResultOption) error {
```
AsIntoto

spencerschrock commented on pkg/scorecard/statement.go:
```diff
@@ -0,0 +1,95 @@
+	Predicate Predicate `json:"predicate"`
+}
+
+// Predicate overrides JSONScorecardResultV2 with a nullable Repo field.
+type Predicate struct {
+	Repo *jsonRepoV2 `json:"repo,omitempty"`
```
for my own understanding, this removes the repo field from the JSON since intoto already has a subject in the statement?

puerco replied:
Correct!

puerco commented:
Oooh sorry I missed the latest reviews! I'm commenting to dismiss the state warning, but I'll handle the review comments tomorrow morning. Thanks!
'

puerco commented:
OK, hopefully, this is the final update that is needed. I've renamed the `IntotoPredicate` and `IntotoPredicateType` variables as suggested above in a39537c02e8e562e674b20dfcfb6cf24d7651642. I've also updated the first commit to pull in the latest intoto/attestaston release.

spencerschrock reviewed (approved):
Thank you!

> OK, hopefully, this is the final update that is needed

And thanks for your patience, this `pkg/scorecard` is the most important API surface for the project as a library, so I wanted to be careful about what was added. And sometimes changes required changes.

puerco commented:
No problem at all, thank you for the thorough review! Happy to get it right :)
